### DISCUSSION TITLE
Can you use axios with a node socket

### DISCUSSION THREAD

**Discussion by Kreijstal** [axios/axios#4893] [Q&A] [answered]
Can I send an http socket over a node stream that I define?

**Comment by jasonsaayman** [comment]
You should be able to, should this not work please put in a PR or open a new issue.

**Reply by Kreijstal** [reply]
My question is how do I tell axios to use a socket instead of allowing it to default to direct tcp streams.
I know this isn't clear so let me do an example..
Say, for example I have another computer with a server that can forward TCP through websockets. So in my computer I have a websocket stream. How do I tell axios through the axios API that I want to send the http request over the websocket, instead of axios making a tcp request directly to the server. I didn't see this in the documentation..

**Comment by jasonsaayman** [comment, marked as answer]
Oh ok, I understand much better now, so no this is not possible as Axios is an HTTP client and not a web socket client. This is also probably not something we will look into doing any time soon as you can use the standard Websocket built in https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_client_applications

**Reply by Kreijstal** [reply]
sure but I just wanted to send the HTTP request as is (Instead of me building the HTTP request by hand), through the websocket client. Imagine HTTP over TCP over Websocket, I guess that's not something you can configure with your library in that case, my question has been answered.

**Reply by jasonsaayman** [reply]
Well if you would like maybe include what the request would look like you can email me and we can try to debug it together. If the request is sent over HTTP then we should actually be able to handle it, if Axios cannot we should for sure be adding it to our list of things to implement.

**Reply by Kreijstal** [reply]
Okay, this is a toy example.
Working example
```js
const { Duplex ,Readable} = require('stream');
const readable = Readable.from(["GET / HTTP/1.1\r\nHost:127.0.0.1\r\n\r\n"])
const fakeTCP = new Duplex({
  write(chunk, encoding, callback) {
	this.strequest+=chunk.toString();
	callback();
	const response="HTTP/1.1 200 OK\r\nContent-Length: 13\r\nConnection: close\r\n\r\nHello, world!";
	if (/^GET \/ HTTP\/1.1\r\nHost:[\S\s]+?\r\n\r\n$/i.test(this.strequest)) {
      this.push(response);
	  this.push(null);
	  this.strequest="";
	}
  },

  read(size) {}
});
fakeTCP.strequest="";
//process.stdin.pipe(fakeTCP).pipe(process.stdout);
readable.pipe(fakeTCP).pipe(process.stdout);
```
possible API
```js
axios.usingStream(fakeTCP).get('/')
```
So basically piping axios through this mock TCP, the reason for this mess is to make a custom proxy where I can push the HTTP messages. (And I didn't want to create real proxies on the OS)
